EIPS/eip-draft_held_token.md

```diff
@@ -0,0 +1,228 @@
+
+## Motivation
+
+As different areas of crypto (DeFi, NFTs, etc.) converge and composability improves, there will more commonly be a distinction between the actual owner (likely a contract) and the functional owner (likely a user) of a token. Currently, this results in a conflict between mechanisms that require token deposits and systems that rely on those tokens for ownership or balance verification.
+
+This proposal aims to address that conflict by providing a standard interface for token holders to expose ownership and balance information. This will allow users to participate in these DeFi mechanisms without giving up existing token utility. Overall, this would greatly increase interoperability across systems, benefiting both users and protocol developers.
+
+Example implementers of this ERC standard include
+- staking or farming contracts
+- lending pools
+- time lock or vesting vaults
+- fractionalized NFT contracts
+
+Example consumers of this ERC standard include
+- governance systems
+- gaming
+- PFP verification
+- art galleries or showcases
+- token based membership programs
+
+## Specification
+
+Smart contracts implementing the held token standard MUST also implement `ERC165` and return true when the corresponding interface ID is passed.
+
+Smart contracts implementing the `ERC20` held token standard MUST implement all of the functions in the `IERC20Holder` interface.
+
+```solidity
+/**
+ * @notice the ERC20 holder standard provides a common interface to query
+ * token balance information
+ */
+interface IERC20Holder is IERC165 {
+  /**
+   * @notice emitted when the token is transferred to the contract
+   * @param owner functional token owner
+   * @param tokenAddress held token address
+   * @param tokenAmount held token amount
+   */
+  event Hold(
+    address indexed owner,
+    address indexed tokenAddress,
+    uint256 tokenAmount
+  );
+
+  /**
+   * @notice emitted when the token is released back to the user
+   * @param owner functional token owner
+   * @param tokenAddress held token address
+   * @param tokenAmount held token amount
+   */
+  event Release(
+    address indexed owner,
+    address indexed tokenAddress,
+    uint256 tokenAmount
+  );
+
+  /**
+   * @notice get the held balance of the token owner
+   * @param tokenAddress held token address
+   * @param owner functional token owner
+   * @return held token balance
+   */
+  function heldBalanceOf(address tokenAddress, address owner)
+    external
+    view
+    returns (uint256);
+}
+
+```
+
+Smart contracts implementing the `ERC721` held token standard MUST implement all of the functions in the `IERC721Holder` interface.
+
+```solidity
+/**
+ * @notice the ERC721 holder standard provides a common interface to query
+ * token ownership and balance information
+ */
+interface IERC721Holder is IERC165 {
+  /**
+   * @notice emitted when the token is transferred to the contract
```

Comment:
Similarly, if a token is transferred with `transferFrom` (instead of `safeTransferFrom`), is the implementer expected to raise the event?

Reply:
Same answer as above, if an ERC721 is sent with the raw `transferFrom` method, the holder contract won't know what to do with it anyways. So this event could be emitted during bookkeeping in either the transfer callback or an explicit deposit method